File: Problem_Solving/leetcode/linear_data_structure/Linked_List/Palindrome_Linked_List_234/palindrome.cpp
/*
    Given the head of a singly linked list, return true if it is a palindrome
    or false otherwise.

    Example 1:

    Input: head = [1,2,2,1]
    Output: true
    Example 2:

    Input: head = [1,2]
    Output: false

    Constraints:

    The number of nodes in the list is in the range [1, 105].
    0 <= Node.val <= 9

    Follow up: Could you do it in O(n) time and O(1) space?
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct ListNode{
    int val;
    ListNode* next;
    // 기본 생성자
    ListNode() : val(0), next(nullptr) {};
    // int 값을 지정하는 생성자. 새로운 리스트 생성할 떄 유용
    ListNode(int x) : val(x), next(nullptr) {};
    // 노드 맨 앞이나 중간 부분에 삽입할 노드 생성에 주로 사용
    ListNode(int x, ListNode* next) : val(x), next(next) {};
};

bool isPalindrome(ListNode *head)
{

}

File: Problem_Solving/leetcode/linear_data_structure/Linked_List/234_Palindrome_Linked_List/palindrome.py
# 리스트 변환

# 팰린드롬 여부를 판별하기 위해 앞 뒤로 모두 추출할 수 있는 자료구조를 사용한다.
def isPalindrome(self, head: ListNode) -> bool:
    q: List = []
    
    if not head:
        return True
    
    node = head
    # 리스트 변환
    while node is not None:
        q.append(node.val)
        node = node.next
        
    # 팰린드롬 판별
    while len(q) > 1:
        # pop 함수에 index 지정이 가능하다.
        if q.pop(0) != q.pop(): # 느리다.
            return False
        
    return True